src/main/java/duke/parser/DukeParser.java

```diff
@@ -0,0 +1,235 @@
+
+        // Split rest of input string for further parsing
+        // If there is no rest of input string, then restOfInputString will be "".
+        this.restOfInputString = newString.substring(this.keyword.length()).trim();
+
+    }
+
+    /**
+     * Executes a loaded and parsed instruction.
+     *
+     * @param st Storage that we want to save data to after the instruction has been executed
+     * @throws DukeException if instruction execution fails
+     */
+    public Command execute() throws DukeException {
+        if (this.keyword == null) {
+            throw new DukeException("Error: Parser has not been loaded with an instruction yet!");
+        }
+
+        if (this.keyword.equals("")) {
+            System.out.println(BREAK_LINES);
+            throw new DukeException("I can't do anything based off a blank instruction!");
+        }
+
+        switch (this.keyword) {
+        case "list":
+            return this.listInstructionHandler();
+            // Intentional Fallthrough
+        case "bye":
+            return this.byeInstructionHandler();
+            // Intentional Fallthrough
+        case "find":
+            return this.findInstructionHandler();
+            // Intentional Fallthrough
+        case "mark":
+            // Intentional Fallthrough
+        case "unmark":
+            // Intentional Fallthrough
+        case "delete":
+            return this.numericalInstructionHandler();
+            // Intentional Fallthrough
+        case "todo":
+            // Intentional fallthrough
+        case "event":
+            // Intentional fallthrough
+        case "deadline":
+            return this.addTaskInstructionHandler();
+            // Intentional Fallthrough
+        default:
+            throw new DukeException("Command not recognised. Try again?");
+        }
+    }
+
+    /**
+     * Sanitises user input.
+     *
+     * @param inputString User input that needs to be sanitised
+     * @return Sanitised user input
+     */
+    public String sanitiseUserInput(String inputString) {
+        // Clear trailing whitespace
+        String out = inputString.trim();
+        return out;
+    }
+
+    /**
+     * Handles a list instruction by printing user's tasks to the screen.
+     */
+    public Command listInstructionHandler() {
+        return new ListCommand();
+    }
+
+    /**
+     * Handles a bye instruction by exiting Duke.
+     */
+    public Command byeInstructionHandler() {
+        return new ByeCommand();
+    }
+
+    /**
+     * Handles numerical instructions that act on an indexed element in taskList
+     *
+     * @throws DukeException Index is invalid
+     */
+    public Command numericalInstructionHandler() throws DukeException {
+
+        // First, try to parse the numerical part of the instruction, if error, throw it
+        int instructionNum;
+        try {
+            instructionNum = Integer.valueOf(this.restOfInputString) - 1;
```

Comment:
Any reason why you used `Integer.valueOf` instead of `parseInt`?

Reply:
Seemed to be the first option that came to mind! Is there any difference in efficiency in using one over the other?

Reply:
`Integer.valueOf` actually uses `parseInt` internally, but it returns a type of Integer rather than int. I just thought for your case, `parseInt` was good enough. No major performance issues if I recall correctly.

```
public static int parseInt(String s) throws NumberFormatException {
    return parseInt(s, 10);
}

public static Integer valueOf(String s, int radix) throws NumberFormatException {
    return Integer.valueOf(parseInt(s, radix));
}

public static Integer valueOf(String s) throws NumberFormatException {
    return Integer.valueOf(parseInt(s, 10));
}
```